server/src/main/java/io/unitycatalog/server/service/FunctionService.java

```diff
@@ -2,43 +2,159 @@
+import lombok.SneakyThrows;
+
+import java.util.List;
+import java.util.Map;
 import java.util.Optional;
+import java.util.UUID;
+
+import static io.unitycatalog.server.model.SecurableType.CATALOG;
+import static io.unitycatalog.server.model.SecurableType.FUNCTION;
+import static io.unitycatalog.server.model.SecurableType.METASTORE;
+import static io.unitycatalog.server.model.SecurableType.SCHEMA;
 
 @ExceptionHandler(GlobalExceptionHandler.class)
 public class FunctionService {
 
   private static final FunctionRepository FUNCTION_REPOSITORY = FunctionRepository.getInstance();
 
-  public FunctionService() {}
+  private static final SchemaRepository SCHEMA_REPOSITORY = SchemaRepository.getInstance();
+  private static final CatalogRepository CATALOG_REPOSITORY = CatalogRepository.getInstance();
+
+  private final UnityCatalogAuthorizer authorizer;
+  private final UnityAccessEvaluator evaluator;
+
+  @SneakyThrows
+  public FunctionService(UnityCatalogAuthorizer authorizer) {
+    this.authorizer = authorizer;
+    evaluator = new UnityAccessEvaluator(authorizer);
+  }
 
   @Post("")
-  public HttpResponse createFunction(CreateFunctionRequest createFunctionRequest) {
-    return HttpResponse.ofJson(FUNCTION_REPOSITORY.createFunction(createFunctionRequest));
+  // TODO: for now, we are not supporting CREATE FUNCTION privilege
+  @AuthorizeExpression("""
+          #authorize(#principal, #metastore, OWNER) ||
+          #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) && #authorizeAny(#principal, #schema, OWNER, USE_SCHEMA)
+          """)
+  @AuthorizeKey(METASTORE)
+  public HttpResponse createFunction(@AuthorizeKeys({
+                                        @AuthorizeKey(value = CATALOG, key = "function_info.catalog_name"),
+                                        @AuthorizeKey(value = SCHEMA, key = "function_info.schema_name")
+                                      })
+                                      CreateFunctionRequest createFunctionRequest) {
+    FunctionInfo functionInfo = FUNCTION_REPOSITORY.createFunction(createFunctionRequest);
+    initializeAuthorizations(functionInfo);
+    return HttpResponse.ofJson(functionInfo);
   }
 
   @Get("")
+  @AuthorizeExpression("#defer")
   public HttpResponse listFunctions(
       @Param("catalog_name") String catalogName,
       @Param("schema_name") String schemaName,
       @Param("max_results") Optional<Integer> maxResults,
       @Param("page_token") Optional<String> pageToken) {
-    return HttpResponse.ofJson(
-        FUNCTION_REPOSITORY.listFunctions(catalogName, schemaName, maxResults, pageToken));
+
+    ListFunctionsResponse listFunctionsResponse = FUNCTION_REPOSITORY.listFunctions(catalogName, schemaName, maxResults, pageToken);
+    filterFunctions("""
+            #authorize(#principal, #metastore, OWNER) ||
+            (#authorizeAll(#principal, #schema, OWNER, USE_SCHEMA) && #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG)) ||
```

Comment:
Isn't `#authorizeAll(#principal, #schema, OWNER, USE_SCHEMA)` the same as `#authorize(#principal, #schema, OWNER)`?

Reply:
Also, once the "catalog owner can list all functions" rule is added, `#authorizeAny(#principal, #catalog, OWNER, USE_CATALOG)` can be reduced to `#authorize(#principal, #catalog, USE_CATALOG)` (also in the next rule)